### PR TITLE
Post mutual requests for external messages

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR resolves the issue with the `IntegrationBus`. Previously, when two repositories were registered sequentially in respective (distinct) contexts, only the first one received the `RequestForExternalMessages` from the second, but not vice-versa. Consequently, only events from the first to second repositories could be dispatched.

Since this PR, `IntegrationBus` sends a `RequestForExternalMessages` when it received such a request from another context. This enables contexts to exchange their required types.

Also, in this PR we tune the in-mem implementation of transport, so that all the memory structures holding references to different context are cleared on `close()`.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: server/src/main/java/io/spine/server/integration/ConfigurationBroadcast.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * Notifies other Bounded contexts about a change in the requested messages.
+     *
+     * <p>If the given {@code needs} are the same as previous ones, i.e. the needs did not change,
+     * the request is not sent.
+     *
+     * @param needs
+     *         the new needs of current context
+     */
+    synchronized void onNeedsUpdated(Set<ChannelId> needs) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/ConfigurationBroadcast.java]: This is unclear. Why would channel IDs be the needs of the Bounded Context? To me, the needs of a context are the types of `Message`s. I.e. "I am `Projects` context and I need `UserCreated` messages".

Could you please rename?

File: server/src/main/java/io/spine/server/integration/ConfigurationChangeObserver.java [hunk 2]
```diff
@@ -27,21 +27,31 @@
...
+    private final IntegrationBus integrationBus;
     private final BoundedContextName boundedContextName;
     private final Function<Class<? extends Message>, BusAdapter<?, ?>> adapterByClass;
 
+    /**
+     * Names of Bounded Contexts already known to this observer.
+     *
+     * <p>If a context is unknown, the observer publishes a {@code RequestForExternalMessages}.
+     */
+    private final Set<BoundedContextName> knownContexts = synchronizedSet(new HashSet<>());
```
Comment (armiol) [inline server/src/main/java/io/spine/server/integration/ConfigurationChangeObserver.java]: Let's define the `hashCode()` for the `BoundedContextName`.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

File: server/src/main/java/io/spine/server/transport/MessageChannel.java [hunk 3]
```diff
@@ -42,4 +39,20 @@
...
+
+    /**
+     * Obtains the type of the messages transferred through this channel.
+     */
+    default TypeUrl targetType() {
+        ChannelId channelId = id();
+        return TypeUrl.parse(channelId.getTargetType());
+    }
+
+    static ChannelId channelIdFor(TypeUrl messageType) {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/transport/MessageChannel.java]: Please document.

**Pull request merged**